vault/router.go

```diff
@@ -457,12 +457,13 @@ func (r *Router) MatchingBackend(ctx context.Context, path string) logical.Backe
 	path = ns.Path + path
 
 	r.l.RLock()
+	defer r.l.RUnlock()
 	_, raw, ok := r.root.LongestPrefix(path)
-	r.l.RUnlock()
 	if !ok {
 		return nil
 	}
-	return raw.(*routeEntry).backend
+	backend := raw.(*routeEntry).backend
+	return backend
```

Comment:
If I understand correctly keeping `return raw.(*routeEntry).backend` in a single line does not solve the race condition because `defer r.l.RUnlock()` is executed right before the return statement, hence the `backend` variable assignation on its own line outside the return statement is necessary?

If yes imo that's worth a comment. I know if I was to go through this code I'd see the "redundant" variable assignation and merge the 2 lines if nothing tells me otherwise.

Reply:
good call!

Reply:
I don't know if assignments behave differently especially since we're dealing with pointers, but defer executions should happen _after_ return calls. Simplified example of defer execution ordering: https://go.dev/play/p/p6bE9H-6j41.

Reply:
I missed the reference on the documentation around the deferred function occurring before the function that's executing returns! I am slightly surprised that the explicit assignment here makes it behave differently, but disregard my last comment if that's what does the trick.

Reply:
@calvn I am actually confused as to why your example behaves differently than mine: https://go.dev/play/p/JSTeicILlfr

Reply:
But yes, from my tests I don't trigger the race detector with that single assignment change in a new statement

Reply:
Following up after doing more tests. The explicit assignment is not necessary after all.